numpy/random/mtrand.pyx

```diff
@@ -3481,16 +3506,20 @@ cdef class RandomState:
             return randoms
 
         _dp = PyFloat_AsDouble(p)
-        _in = <long>n
+        _in = int(n)
         check_constraint(_dp, 'p', CONS_BOUNDED_0_1)
```

Comment:
Is the a python function call rather than a C cast? Maybe Cython 3 does the right thing here?

Reply:
The problem was that this can be a Python float, somehow the `<long>` cast uses a long cast that allows floats, while other to integer casts don't.  This _does_ depend on the Cython version (it worked locally).

I suspect it's the opposite, it was necessary in Cython 3 to do something more than a cast.

Will try once more locally, but maybe the best path is to just extract the value from the array (which we have created here in either case).

Reply:
Could you just case twice? `<long>(<int>n)` to avoid python?